core/controller/src/main/scala/whisk/core/controller/Meta.scala

```diff
@@ -190,77 +448,108 @@ trait WhiskMetaApi extends Directives with PostActionActivation {
-                // fail the request with MethodNotAllowed so as not to leak information about the existence
-                // of packages that are otherwise private
-                info(this, s"meta api request references package which is missing")
-                Future.failed(RejectRequest(MethodNotAllowed))
-        } flatMap { pkg =>
+    private def confirmMetaPackage(pkgLookup: Future[WhiskPackage], method: HttpMethod)(
+        implicit transid: TransactionId): Future[(EntityName, Parameters)] = {
+        pkgLookup flatMap { pkg =>
             // expecting the meta handlers to be private; should it be an error? warn for now
             if (pkg.publish) {
                 warn(this, s"'${pkg.fullyQualifiedName(true)}' is public")
             }
 
-            if (pkg.binding.isEmpty) {
-                pkg.annotations.get("meta") filter {
-                    // does package have annotation: meta == true
-                    _ match { case JsBoolean(b) => b case _ => false }
-                } flatMap {
-                    // if so, find action name for http verb
-                    _ => pkg.annotations.get(method.name.toLowerCase)
-                } match {
-                    // if action name is defined as a string, accept it, else fail request
-                    case Some(JsString(actionName)) =>
-                        info(this, s"'${pkg.name}' maps '${method.name}' to action '${actionName}'")
-                        Future.successful(EntityName(actionName), pkg.parameters)
-                    case _ =>
-                        info(this, s"'${pkg.name}' is missing 'meta' annotation or action name for '${method.name.toLowerCase}'")
-                        Future.failed(RejectRequest(MethodNotAllowed))
-                }
-            } else {
-                warn(this, s"'${pkg.fullyQualifiedName(true)}' is a binding")
-                Future.failed(RejectRequest(MethodNotAllowed))
+            // does package have annotation: meta == true and a
+            // mapping from HTTP verb to action name?
+            val isMetaPackage = pkg.annotations.asBool("meta")
+
+            isMetaPackage flatMap {
+                _ => pkg.annotations.asString(method.name.toLowerCase)
+            } match {
+                // if action name is defined as a string, accept it, else fail request
+                case Some(actionName) =>
+                    info(this, s"'${pkg.name}' maps '${method.name}' to action '${actionName}'")
+                    Future.successful(EntityName(actionName), pkg.parameters)
+
+                case _ =>
+                    info(this, s"'${pkg.name}' is missing 'meta' annotation or action name for '${method.name.toLowerCase}'")
+                    isMetaPackage match {
+                        case Some(true) => Future.failed(RejectRequest(MethodNotAllowed))
+                        case _          => Future.failed(RejectRequest(NotFound))
+                    }
```

Comment:
Same as below on converting `Option[Boolean]` to `Boolean`, handling `None` as `false`:

```
val isMetaPackage = pkg.annotations.asBool("meta").exists(identity)
if(isMetaPackage) {
    pkg.annotations.asString(method.name.toLowerCase).map { actionName => 
        info(this, s"'${pkg.name}' maps '${method.name}' to action '${actionName}'")
        Future.successful(EntityName(actionName), pkg.parameters)
    }.getOrElse {
        info(this, s"'${pkg.name}' is missing action name for '${method.name.toLowerCase}'")
        Future.failed(RejectRequest(MethodNotAllowed))
    }
} else {
    info(this, s"'${pkg.name}' is missing 'meta' annotation")
    Future.failed(RejectRequest(NotFound))
}
```

To me that is easier to follow what exactly is happening and as a bonus, we get a disambiguated log message.